beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -49,81 +49,65 @@ func (s *Service) processPendingBlocksQueue() {
 	})
 }
 
-// processes the block tree inside the queue
+// processPendingBlocks validates, processes, and broadcasts pending blocks.
 func (s *Service) processPendingBlocks(ctx context.Context) error {
 	ctx, span := trace.StartSpan(ctx, "processPendingBlocks")
 	defer span.End()
 
-	pids := s.cfg.p2p.Peers().Connected()
+	// Validate pending slots before processing.
 	if err := s.validatePendingSlots(); err != nil {
 		return errors.Wrap(err, "could not validate pending slots")
 	}
-	ss := s.sortedPendingSlots()
-	var parentRoots [][32]byte
 
-	span.AddAttributes(
-		trace.Int64Attribute("numSlots", int64(len(ss))),
-		trace.Int64Attribute("numPeers", int64(len(pids))),
-	)
+	// Sort slots for ordered processing.
+	sortedSlots := s.sortedPendingSlots()
+
+	span.AddAttributes(trace.Int64Attribute("numSlots", int64(len(sortedSlots))), trace.Int64Attribute("numPeers", int64(len(s.cfg.p2p.Peers().Connected()))))
 
 	randGen := rand.NewGenerator()
-	for _, slot := range ss {
-		// process the blocks during their respective slot.
-		// otherwise wait for the right slot to process the block.
-		if slot > s.cfg.clock.CurrentSlot() {
+	var parentRoots [][32]byte
+
+	// Iterate through sorted slots.
+	for _, slot := range sortedSlots {
+		// Skip processing if current slot is ahead.
+		if shouldSkipProcessing(slot, s.cfg.clock.CurrentSlot()) {
 			continue
 		}
 
-		ctx, span := trace.StartSpan(ctx, "processPendingBlocks.InnerLoop")
-		span.AddAttributes(trace.Int64Attribute("slot", int64(slot))) // lint:ignore uintcast -- This conversion is OK for tracing.
+		ctx, span := startInnerSpan(ctx, slot)
 
-		s.pendingQueueLock.RLock()
-		bs := s.pendingBlocksInCache(slot)
-		// Skip if there's no block in the queue.
-		if len(bs) == 0 {
-			s.pendingQueueLock.RUnlock()
+		// Get blocks in the pending queue for the current slot.
+		blocksInCache := s.getBlocksInQueue(slot)
+		if len(blocksInCache) == 0 {
 			span.End()
 			continue
 		}
-		s.pendingQueueLock.RUnlock()
 
-		// Loop through the pending queue and mark the potential parent blocks as seen.
-		for _, b := range bs {
-			if b == nil || b.IsNil() || b.Block().IsNil() {
-				span.End()
+		// Process each block in the queue.
+		for _, b := range blocksInCache {
+			if shouldSkipBlock(b) {
```

Comment:
Could we use this instead? https://github.com/prysmaticlabs/prysm/blob/develop/consensus-types/blocks/getters.go#L22